src/__test__/test_production_3.py

```diff
@@ -0,0 +1,163 @@
+import unittest
+import itertools as it
+
+from graph import Graph
+from model import Node, NodeAttrs, EdgeAttrs, Edge
+
+from production import P3
+
+class TestProduction2(unittest.TestCase):
+    def test_production_can_be_applied_on_lhs_graph(self):
+        graph = Graph()
+
+        node_0 = Node(NodeAttrs('v', 0, 0, False))
+        node_1 = Node(NodeAttrs('v', 1, 0, False))
+        node_2 = Node(NodeAttrs('v', 1, 1, False))
+        node_3 = Node(NodeAttrs('v', 0, 1, False))
+        node_4 = Node(NodeAttrs('v', 1, 0.5, True))
+        node_5 = Node(NodeAttrs('v', 0.5, 0, True))
+        nodes = [node_0, node_5, node_1, node_4, node_2, node_3]
+
+        graph.add_node_collection(nodes)
+        graph.add_q_hyperedge((node_0, node_1, node_2, node_3), EdgeAttrs('q', True))
+
+        for node_1, node_2 in it.pairwise(nodes + [node_0]):
+            graph.add_edge(Edge(node_1.handle, node_2.handle, EdgeAttrs(kind='e', flag=False)))
+
+        p3 = P3()
+        mapping_gen = graph.generate_subgraphs_isomorphic_with(p3.get_lhs())
+        self.assertTrue(any(p3.is_mapping_feasible(graph, mapping) for mapping in mapping_gen))
+
+    def test_production_can_be_applied_if_hanging_nodes_in_different_places(self):
```

Comment:
These tests are effectively useless, as `is_mapping_feasible` returns `True` by default unless overridden

We should rather test whether graph structure after applying the production looks like expected.

I believe you just followed other unittests, but they are also useless after #17 was merged, thus either remove them completely or let's made them actually do something.